### ISSUE TITLE
Issue when chaining pvagw

### ISSUE THREAD

**Issue by beenje** [mdavidsaver/p4p#53]
Due to isolation purpose we sometimes have to chain EPICS gateways.
I haven't seen any issue with cagw. With pvagw, the result isn't stable.
If I do client -> pvagw02 -> IOC, everything works fine.
I I do client -> pvagw01 -> pvagw02 -> IOC, the first pvget request works properly, but the second one timeout. If I wait a while, it works again. Repetitive requests timeout.

```
(epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:52:31 CEST 2020
demo:bb-IocStats:MEM_USED 2020-07-09 16:52:26.663  1.26525e+07
(epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:52:32 CEST 2020
Timeout
demo:bb-IocStats:MEM_USED (epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:52:39 CEST 2020
Timeout
demo:bb-IocStats:MEM_USED (epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:52:45 CEST 2020
Timeout
demo:bb-IocStats:MEM_USED (epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:52:51 CEST 2020
Timeout
demo:bb-IocStats:MEM_USED (epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:53:08 CEST 2020
Timeout
demo:bb-IocStats:MEM_USED (epics) CI0011906:~ $ date; pvget demo:bb-IocStats:MEM_USED
Thu Jul  9 16:53:35 CEST 2020
demo:bb-IocStats:MEM_USED 2020-07-09 16:53:26.622  1.26525e+07
```

The pvagw01 addrlist just points to the pvagw02 server interface.

Is this a scenario that should work properly?
Both pvagw are deployed with p4p 3.4.2 and epics-base 7.0.3.1.

**Comment by mdavidsaver** [comment]
This is strange.  (and yes, it should work)  The ~60 sec time matches up with the channel cache cleanup interval.

Could you test with `pvmonitor` and/or set a much longer timeout with `pvget`?  eg. `-w 1000`.

In addition to enabling debug logging with `-v` (cf. #52) you might also query the channel cache info of the two gateways with [\<statusprefix\>cache](https://mdavidsaver.github.io/p4p/gw.html#status-pvs).

**Comment by beenje** [comment]
Using pvmonitor works fine:
```
$ pvmonitor demo:bb:counter
demo:bb:counter 2020-07-10 11:26:20.472  8
demo:bb:counter 2020-07-10 11:26:22.472  9
demo:bb:counter 2020-07-10 11:26:24.471  10
demo:bb:counter 2020-07-10 11:26:26.471  0
demo:bb:counter 2020-07-10 11:26:28.470  1
```

Using a longer timeout doesn't help:
```
$ date; pvget -w 65 demo:bb:counter
Fri Jul 10 11:30:17 CEST 2020
demo:bb:counter 2020-07-10 11:30:16.470  5
$ date; pvget -w 65 demo:bb:counter
Fri Jul 10 11:30:21 CEST 2020
Timeout
```

With `-d`:
```
$ date; pvget -w 10 -d demo:bb:counter
Fri Jul 10 11:36:51 CEST 2020
2020-07-10T11:36:51.103 Creating datagram socket from: 0.0.0.0:60779.
2020-07-10T11:36:51.104 Broadcast address #0: 172.30.8.12:5076. (unicast)
2020-07-10T11:36:51.104 Setting up UDP for interface 192.168.75.117/255.255.255.0, broadcast 192.168.75.255, dest <none>.
2020-07-10T11:36:51.104 Creating datagram socket from: 192.168.75.117:5076.
2020-07-10T11:36:51.104 Creating datagram socket from: 192.168.75.255:5076.
2020-07-10T11:36:51.104 Setting up UDP for interface 172.18.157.59/255.255.255.255, broadcast 172.18.157.59, dest <none>.
2020-07-10T11:36:51.104 Creating datagram socket from: 172.18.157.59:5076.
2020-07-10T11:36:51.104 Unable to find broadcast address of interface 172.18.157.59.
2020-07-10T11:36:51.104 Creating datagram socket from: 224.0.0.128:5076.
2020-07-10T11:36:51.104 Local multicast enabled on 127.0.0.1/224.0.0.128:5076.
2020-07-10T11:36:51.104 Sending 61 bytes 0.0.0.0:60779 -> 172.30.8.12:5076.
Waiting...
2020-07-10T11:36:51.339 Sending 61 bytes 0.0.0.0:60779 -> 172.30.8.12:5076.
2020-07-10T11:36:51.358 UDP Client Rx (53) 0.0.0.0:60779 <- 172.30.8.12:45918
2020-07-10T11:36:51.359 Connecting to PVA server: 172.30.8.12:5075.
2020-07-10T11:36:51.359 Opening socket to PVA server 172.30.8.12:5075, attempt 1.
2020-07-10T11:36:51.374 Socket connected to PVA server: 172.30.8.12:5075.
2020-07-10T11:36:51.374 Acquiring transport to 172.30.8.12:5075.
2020-07-10T11:36:51.414 Connected to PVA server: 172.30.8.12:5075.
demo:bb:counter 2020-07-10 11:36:50.471  4
2020-07-10T11:36:51.480 Releasing TCP transport to 172.30.8.12:5075.
2020-07-10T11:36:51.481 TCP socket to 172.30.8.12:5075 is to be closed.
2020-07-10T11:36:51.481 UDP socket 0.0.0.0:0 closed.
2020-07-10T11:36:51.481 UDP socket 192.168.75.117:5076 closed.
2020-07-10T11:36:51.481 UDP socket 192.168.75.255:5076 closed.
2020-07-10T11:36:51.481 UDP socket 172.18.157.59:5076 closed.
2020-07-10T11:36:51.481 UDP socket 224.0.0.128:5076 closed.
Done

$ date; pvget -w 10 -d demo:bb:counter
Fri Jul 10 11:37:14 CEST 2020
2020-07-10T11:37:14.426 Creating datagram socket from: 0.0.0.0:58227.
2020-07-10T11:37:14.427 Broadcast address #0: 172.30.8.12:5076. (unicast)
2020-07-10T11:37:14.427 Setting up UDP for interface 192.168.75.117/255.255.255.0, broadcast 192.168.75.255, dest <none>.
2020-07-10T11:37:14.427 Creating datagram socket from: 192.168.75.117:5076.
2020-07-10T11:37:14.427 Creating datagram socket from: 192.168.75.255:5076.
2020-07-10T11:37:14.427 Setting up UDP for interface 172.18.157.59/255.255.255.255, broadcast 172.18.157.59, dest <none>.
2020-07-10T11:37:14.427 Creating datagram socket from: 172.18.157.59:5076.
2020-07-10T11:37:14.427 Unable to find broadcast address of interface 172.18.157.59.
2020-07-10T11:37:14.427 Creating datagram socket from: 224.0.0.128:5076.
2020-07-10T11:37:14.427 Local multicast enabled on 127.0.0.1/224.0.0.128:5076.
2020-07-10T11:37:14.428 Sending 61 bytes 0.0.0.0:58227 -> 172.30.8.12:5076.
Waiting...
2020-07-10T11:37:14.493 UDP Client Rx (53) 0.0.0.0:58227 <- 172.30.8.12:45918
2020-07-10T11:37:14.493 Connecting to PVA server: 172.30.8.12:5075.
2020-07-10T11:37:14.493 Opening socket to PVA server 172.30.8.12:5075, attempt 1.
2020-07-10T11:37:14.509 Socket connected to PVA server: 172.30.8.12:5075.
2020-07-10T11:37:14.509 Acquiring transport to 172.30.8.12:5075.
2020-07-10T11:37:14.558 Connected to PVA server: 172.30.8.12:5075.
Timeout
demo:bb:counter 2020-07-10T11:37:24.432 Releasing TCP transport to 172.30.8.12:5075.
2020-07-10T11:37:24.432 TCP socket to 172.30.8.12:5075 is to be closed.
2020-07-10T11:37:24.432 UDP socket 0.0.0.0:0 closed.
2020-07-10T11:37:24.432 UDP socket 192.168.75.117:5076 closed.
2020-07-10T11:37:24.432 UDP socket 192.168.75.255:5076 closed.
2020-07-10T11:37:24.432 UDP socket 172.18.157.59:5076 closed.
2020-07-10T11:37:24.432 UDP socket 224.0.0.128:5076 closed.
Done
```

On the pvagw side:
```
DEBUG:p4p.gw:172.30.8.12:5076 Searching for b'demo:bb:counter'
DEBUG:p4p.gw:allowed: b'demo:bb:counter' by 172.30.8.12:5076 -> 0
DEBUG:p4p.gw:172.30.8.12:5076 Searching for b'demo:bb:counter'
DEBUG:p4p.gw:allowed: b'demo:bb:counter' by 172.30.8.12:5076 -> 1
DEBUG:p4p.gw:Create b'demo:bb:counter' by 172.18.157.59:50007
DEBUG:p4p.asLib:(re)create <p4p._gw.Channel object at 0x7f70e1674830>, DEFAULT, benjaminbertrand, 172.18.157.59, 0
DEBUG:p4p.gw:172.30.8.12:5076 Searching for b'demo:bb:counter'
DEBUG:p4p.gw:allowed: b'demo:bb:counter' by 172.30.8.12:5076 -> 1
DEBUG:p4p.gw:Create b'demo:bb:counter' by 172.18.157.59:50012
DEBUG:p4p.asLib:(re)create <p4p._gw.Channel object at 0x7f70dabdb7f0>, DEFAULT, benjaminbertrand, 172.18.157.59, 0
```

Here is an example with more traces: https://gist.github.com/beenje/0cc8353561dd8931dd6bd3ae61d53d1d

And yes, the command seems to work again only after a "Channel Cache sweep" on the gw.

**Comment by beenje** [comment]
Hello @mdavidsaver,
Any feedback on this issue? Is there anything else I could try?

**Comment by juanfem** [comment]
Hello, any news on this? I am an user of the gateways @beenje mentioned and I am experiencing this issue...

**Comment by andersharrisson-ess** [comment]
Hi @mdavidsaver ,
This is still an issue and it can be reproduced in this Docker environment:
```
$ git clone -q https://gitlab.esss.lu.se/andersharrisson/epics-gateway-test.git
$ cd epics-gateway-test/
$ docker-compose up
Starting epics-gateway-test_ioc_1            ... done
Starting epics-gateway-test_ioc-gateway_1    ... done
Starting epics-gateway-test_client-gateway_1 ... done
Starting epics-gateway-test_client_1         ... done
...
...
client_1          | Alive
client_1          | test:random 2020-11-23 10:05:16.388  0.750072 
client_1          | Alive
client_1          | Timeout
```
Let me know if I can help to further debug this issue.

**Comment by mdavidsaver** [comment]
I apologize for ignoring this issue for so long.  It hasn't been high on my priority list.  If I'm being honest, this is in part because it looks like it will be a real pain to sort out.

I'm not sure how useful the docker config will be to me.  I believe you that it demonstrates the problem.  However, I doubt this (likely minimal) environment will be enough for me to troubleshoot in (eg. wireshark, gdb, and/or rebuilding with extra prints).

I'm not sure how much is reasonable to ask you all to do.  It would be easiest for me for me if this issue could be replicated on a single host (aka. my laptop), with the three different network segments modeled with different port numbers (like [loopback.conf](https://github.com/mdavidsaver/p4p/blob/master/loopback.conf)).

**Comment by andersharrisson-ess** [comment]
Hi,

Yes this should be reproducible without Docker as well.

Start two gateways:
```
$ cat > /tmp/gw.conf
{
    "version":2,
    "readOnly":true,
    "clients":[
        {
            "name":"gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5086
        },
        {
            "name":"chained-gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5096
        }
    ],
    "servers":[
        {
            "name":"gateway",
            "clients":["gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5075,
            "bcastport":5076,
            "statusprefix":"gateway:"
        },
        {
            "name":"chained-gateway",
            "clients":["chained-gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5085,
            "bcastport":5086,
            "statusprefix":"chained-gateway:"
        }
    ]
}
```
Start a server reachable from the chained-gateway client:
```
$ EPICS_PVA_BROADCAST_PORT=5096 EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO python -m p4p.server.cli my:pv:name=int
INFO:__main__:Add my:pv:name
2021-01-07T14:19:31.855 Using dynamically assigned TCP port 36721.
INFO:p4p.server:Running server
```
Try to get the value with a client:
```
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO python -m p4p.client.cli get my:pv:name
my:pv:name Thu Jan  7 14:20:47 2021 0
```
Try to get the value again:
```
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO python -m p4p.client.cli get my:pv:name
my:pv:name Error: 
```
Monitoring still works:
```
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO python -m p4p.client.cli monitor my:pv:name
my:pv:name Error: Channel disconnected
my:pv:name Thu Jan  7 14:20:47 2021 0
```
I hope this will help resolve the issue.

**Comment by andersharrisson-ess** [comment]
Hi @mdavidsaver ,

Do you have any update on this issue? Let me know if there is anything more we can do to help this ticket along.

**Comment by mdavidsaver** [comment]
On #54 I've added a test case for chaining two gateways, which is passing.  While #54 is still under development, I think it has reached the point of being testable by someone at ess.

**Comment by andersharrisson-ess** [comment]
Thank you for the status update, @mdavidsaver . :+1:

**Comment by andersharrisson-ess** [comment]
I am still unable to successfully chain the gateways. In fact, with pvxs I am not able to retrieve the PV at all. These are the steps I use to run my test:

```
$ pip install p4p==4.0.0a1 pvxslibs  # Installing p4p and pvxs
$ cat > /tmp/conf.json << EOF
{
    "version":2,
    "readOnly":true,
    "clients":[
        {
            "name":"gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5086
        },
        {
            "name":"chained-gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5096
        }
    ],
    "servers":[
        {
            "name":"gateway",
            "clients":["gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5075,
            "bcastport":5076,
            "statusprefix":"gateway:"
        },
        {
            "name":"chained-gateway",
            "clients":["chained-gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5085,
            "bcastport":5086,
            "statusprefix":"chained-gateway:"
        }
    ]
}
EOF
$ pvagw /tmp/conf.json &  # Starting two gateways on separate ports
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO EPICS_PVA_BROADCAST_PORT=5096 python -m p4p.client.cli get my:pv:name # Test connecting directly to the server - Successful!
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO EPICS_PVA_BROADCAST_PORT=5086 python -m p4p.client.cli get my:pv:name # Test connecting through one gateway - Successful!
$ EPICS_PVA_ADDR_LIST=127.0.0.1 EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 EPICS_PVA_AUTO_ADDR_LIST=NO EPICS_PVA_BROADCAST_PORT=5076 python -m p4p.client.cli get my:pv:name # Test connecting through two chained gateways - Fails!
```
Have I misunderstood something with the implementation? I would appreciate any pointer in the right direction.

**Comment by mdavidsaver** [comment]
In this instance the issue is only the automatic self-loop prevention causing the the `"gateway"` client it ignore responses from the `"changed-gateway"` server.  If I split your example configuration into two separate gateway processes then I am able to GET and MONITOR through both.

```
$ cat ess1.conf 
{
    "version":2,
    "readOnly":true,
    "clients":[
        {
            "name":"gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5086
        }
    ],
    "servers":[
        {
            "name":"gateway",
            "clients":["gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5075,
            "bcastport":5076,
            "statusprefix":"gateway:"
        }
    ]
}
$ cat ess2.conf 
{
    "version":2,
    "readOnly":true,
    "clients":[
        {
            "name":"chained-gateway-client",
            "provider":"pva",
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "bcastport":5096
        }
    ],
    "servers":[
        {
            "name":"chained-gateway",
            "clients":["chained-gateway-client"],
            "interface":["127.0.0.1"],
            "addrlist":"127.0.0.1",
            "autoaddrlist":false,
            "serverport":5085,
            "bcastport":5086,
            "statusprefix":"chained-gateway:"
        }
    ]
}
```

```
$ ./bin/linux-x86_64-debug/pvagw ess1.conf
```

```
$ ./bin/linux-x86_64-debug/pvagw ess2.conf
```

```
$ EPICS_PVA_ADDR_LIST=127.0.0.1 \
  EPICS_PVAS_INTF_ADDR_LIST=127.0.0.1 \
  EPICS_PVA_AUTO_ADDR_LIST=NO \
  EPICS_PVA_BROADCAST_PORT=5076 \
  ./bin/linux-x86_64/pvxmonitor cnt
cnt Connected to 127.0.0.1:41873
cnt
value double = 42
alarm struct "alarm_t"
alarm.severity int32_t = 0
alarm.status int32_t = 0
alarm.message string = ""
timeStamp struct "time_t"
timeStamp.secondsPastEpoch int64_t = 0
timeStamp.nanoseconds int32_t = 0
timeStamp.userTag int32_t = 0
```

I also tried restarting the three PVA servers in different orders, and end client reconnected in all cases.

**Comment by andersharrisson-ess** [comment]
Indeed, it works fine when they run in different processes. Thank you for the information, @mdavidsaver.
I had been running the tests with "--no-ban-local", which worked for p4p 3.x (except for the issue stated in this topic). So what role does "--no-ban-local" play in p4p 4.x?

**Comment by mdavidsaver** [comment]
I saw `--no-ban-local` as a crude but necessary hack.  The pvAccessCPP library API doesn't provide enough information for to reliably distinguish a self-loop from eg. a `pvget` originating from the gateway host.  PVXS does, by providing the ([semi-public](https://mdavidsaver.github.io/pvxs/details.html#expert-apis)) `pvxs::client::Context::ignoreServerGUIDs()` method.

With 4.x, an instance will configure all GW clients with the GUIDs of associated GW servers so that any response from GW server to GW client will be ignored.  I think this will provide good, and fully automatic, protection against self loops within a single gateway process.  This also simplifies testing with local `pvget`.

Passing `--no-ban-local` to `4.0.0a1` now prints a warning.

https://github.com/mdavidsaver/p4p/blob/e2163b11a1a19151e69f712537fddf6d71717b31/src/p4p/gw.py#L698

The documentation will be updated to mention this deprecation prior to `4.0.0`.

**Comment by andersharrisson-ess** [comment]
Thank you for the clarification.
We anxiously await the release of 4.x. :+1:

**Comment by mdavidsaver** [comment]
Fixed w/ 4.0.0
